### PR TITLE
Feature/sjs perf

### PR DESCRIPTION
**Pull request by paxtonhare**
enjoy!

### CODE REVIEW

File: examples/performance-sample/build.gradle [hunk 1]
```diff
@@ -1,8 +1,20 @@
...
+  }
+  dependencies {
+    classpath "com.marklogic:ml-data-hub:2.0.5"
+  }
+}
+
 plugins {
     id 'net.saliman.properties' version '1.4.6'
-    id 'com.marklogic.ml-data-hub' version '2.0.4'
+    // id 'com.marklogic.ml-data-hub' version '2.0.4'
```
Comment (dmcassel) [inline examples/performance-sample/build.gradle]: delete?

File: examples/performance-sample/build.gradle [hunk 2]
```diff
@@ -417,3 +433,110 @@ class TimingsListener implements TaskExecutionListener, BuildListener {
...
+import com.marklogic.client.DatabaseClient
+import com.marklogic.client.io.FileHandle
+import com.marklogic.contentpump.bean.MlcpBean
+import org.gradle.api.logging.Logger
+import org.gradle.api.logging.Logging
+import org.gradle.api.tasks.JavaExec
+import org.gradle.api.tasks.TaskAction
+
+import com.marklogic.appdeployer.AppConfig
+class MlcpTask extends JavaExec {
```
Comment (dmcassel) [inline examples/performance-sample/build.gradle]: isn't this part of ml-gradle? from [Rob's wiki](https://github.com/marklogic-community/ml-gradle/wiki/Content-Pump-and-Gradle) I wouldn't think this would be needed (but I'm still learning about gradle)

File: marklogic-data-hub/src/server-side/dhf.sjs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    tracelib.setPluginLabel(label);
+  } else {
+    fn.error(xs.QName("CONTEXT_MISSING_LABEL"), "Your context object is missing a label");
+  }
+
+  tracelib.resetPluginInput();
+
+  let inputs = context.inputs
+  for (let key in inputs) {
+    tracelib.setPluginInput(key, inputs[key]);
```
Comment (dmcassel) [inline marklogic-data-hub/src/server-side/dhf.sjs]: ``if (obj.hasOwnProperty(prop)) {`` needed here?

File: marklogic-data-hub/src/server-side/dhf.sjs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+  tracelib.pluginTrace(null, "PT0S");
+};
+
+function dhfRequire(moduleUri, func) {
+  if (_requireCache[moduleUri]) {
+    return _requireCache[moduleUri];
+  }
+
+  let cache = func();
+  //require.call(caller, moduleUri);
```
Comment (dmcassel) [inline marklogic-data-hub/src/server-side/dhf.sjs]: remove commented out line

File: examples/performance-sample/build.gradle (hunk 1 above)
Comment (paxtonhare) [reply examples/performance-sample/build.gradle]: should not have checked in.. will fix

File: examples/performance-sample/build.gradle (hunk 2 above)
Comment (paxtonhare) [reply examples/performance-sample/build.gradle]: he hasn't updated to the latest one yet. I forgot to put a comment in there about it. will do that.

File: marklogic-data-hub/src/server-side/impl/trace-lib.sjs [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+        if (value instanceof BinaryNode) {
+          value = "binary data";
+        }
+        oo[key] = value;
+      }
+      return oo;
+    }
+    else {
+      const x = new NodeBuilder();
+      return x.toNode();
```
Comment (dmcassel) [inline marklogic-data-hub/src/server-side/impl/trace-lib.sjs]: returning here means the following lines won't be run. Looks like `return x.toNode();` should be after the for loop.

File: marklogic-data-hub/src/server-side/impl/trace-lib.sjs [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+        }
+      };
+      let i;
+      for (i = 0; i < currentTrace.traceSteps.length; i++) {
+        let step = currentTrace.traceSteps[i];
+        trace.trace.steps.push(step);
+      }
+    }
+    else {
+      const x = new NodeBuilder();
```
Comment (dmcassel) [inline marklogic-data-hub/src/server-side/impl/trace-lib.sjs]: would prefer `nb` to `x`

File: marklogic-data-hub/src/server-side/impl/trace-lib.sjs [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+    addFailedItem(identifier);
+  };
+  let ts = getCurrentTraceSettings();
+  ts['_has_errors'] = true;
+  let traceSteps = currentTrace.traceSteps || [];
+
+  traceSteps.push({
+    label: getPluginLabel(currentTrace),
+    input: getPluginInput(currentTrace),
+    error: rfc.isJson() ? error : error,
```
Comment (dmcassel) [inline marklogic-data-hub/src/server-side/impl/trace-lib.sjs]: if pushing the same thing either way, why the test?

File: marklogic-data-hub/src/server-side/impl/trace-lib.sjs [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+
+function getTraces(page, pageLength) {
+  return tracelib.getTraces(page, pageLength);
+}
+
+function getTrace(id) {
+  return tracelib.getTrace(id);
+}
+
+function getTraceIds(q) {
```
Comment (dmcassel) [inline marklogic-data-hub/src/server-side/impl/trace-lib.sjs]: is `q` a query? Either a comment or a better label would be useful

File: marklogic-data-hub/src/test/java/com/marklogic/hub/EndToEndFlowTests.java [hunk 9]
```diff
@@ -187,6 +208,7 @@ private JsonNode validateUserModules() {
         DataHub dataHub = getDataHub();
         List<DynamicTest> tests = new ArrayList<>();
 
+        // flows before a structured change get updated. this ru
```
Comment (dmcassel) [inline marklogic-data-hub/src/test/java/com/marklogic/hub/EndToEndFlowTests.java]: comment missing part of the sentence

File: marklogic-data-hub/src/test/java/com/marklogic/hub/EndToEndFlowTests.java [hunk 10]
```diff
@@ -413,10 +435,13 @@ private JsonNode validateUserModules() {
                     Map<String, Object> options = new HashMap<>();
                     options.put(plugin + "GoBoom", true);
                     if(!isSslRun() && !isCertAuth()) {
-                        tests.add(DynamicTest.dynamicTest(flowName + ": " + plugin + " error MLCP", () -> {
-                            FinalCounts finalCounts = new FinalCounts(0, 0, 1, 1, 0, 0, 0, 0, 0, 0, "FAILED");
-                            testInputFlowViaMlcp(prefix, "-2", stagingClient, codeFormat, dataFormat, useEs, options, finalCounts);
-                        }));
+//                    tests.add(DynamicTest.dynamicTest(flowName + ": " + plugin + " error MLCP", () -> {
```
Comment (dmcassel) [inline marklogic-data-hub/src/test/java/com/marklogic/hub/EndToEndFlowTests.java]: remove outdated code

File: marklogic-data-hub/src/server-side/impl/trace-lib.sjs (hunk 5 above)
Comment (paxtonhare) [reply marklogic-data-hub/src/server-side/impl/trace-lib.sjs]: looks like an edge case that would break if it were properly tested. I'll fix that.

File: marklogic-data-hub/src/test/java/com/marklogic/hub/EndToEndFlowTests.java (hunk 9 above)
Comment (paxtonhare) [reply marklogic-data-hub/src/test/java/com/marklogic/hub/EndToEndFlowTests.java]: I'm going to remove it because I have no clue what I was typing there.

File: marklogic-data-hub/src/server-side/dhf.sjs (hunk 3 above)
Comment (paxtonhare) [reply marklogic-data-hub/src/server-side/dhf.sjs]: nope. that blows up.

Comment (dmcassel) [reply marklogic-data-hub/src/server-side/dhf.sjs]: I meant inputs.hasOwnProperty. Only relevant if inputs might inherit properties from somewhere.

Comment (paxtonhare) [reply marklogic-data-hub/src/server-side/dhf.sjs]: yeah. I fixed it to the right thing and it still blew up.
